### PR TITLE
lib: os: Rename icmsg_buf to spsc_pbuf

### PR DESCRIPTION
**Pull request by nordic-krch**
Move icmsg_buf to lib/os and rename to spsc_pbuf (Single Producer
Single Consumer Packet Buffer). It is a generic module and initially
was created as internal module for ipc service.

### CODE REVIEW

**Comment by nordic-krch** [comment]
@emob-nordic @pdunaj I would like to use this buffer elsewhere and soon extend it as proposed in https://github.com/zephyrproject-rtos/zephyr/pull/43544#issuecomment-1070755803

**Comment by carlocaione** [comment]
Can you rename also the IPC service backend?

**Review by pdunaj — DISMISSED** [review]

File: tests/lib/spsc_pbuf/src/main.c [hunk 1]
```diff
@@ -5,45 +5,45 @@
...
  */
 static uint8_t memory_area[212] __aligned(4);
 
-static void test_icmsg_buf_ut(void)
+static void test_spsc_pbuf_ut(void)
 {
 	static uint8_t rbuf[198];
 	static uint8_t message[20] = {'a'};
-	struct icmsg_buf *ib;
+	struct spsc_pbuf *ib;
```
Comment (emob-nordic) [inline tests/lib/spsc_pbuf/src/main.c:20]: Please also change ib -> pbuf (packet buffer) or similar. All across the file and in similar places.

**Review by eobalski — CHANGES_REQUESTED** [review]
The short name is a bit complicated, but i don't see any better suggestion. Maybe just circular packet buffer instead (`cpb`)?

I nacked because of small fixes commented inline.

**Comment by nordic-krch** [comment]
regarding renaming of ipc service. I wonder if it should be renamed to something that would indicate its feature (lightweight, simplicity). Was thinking about something like `ipc_lite`? That will tell more to the new user than `ipc_spsc_pbuf` which tells nothing.

**Comment by nordic-krch** [comment]
^ @pdunaj @emob-nordic @carlocaione

**Comment by carlocaione** [comment]
`ipc_lite` is way too generic IMO (what if in the future we add a new backend even lighter?). If the the service is actually using `spsc_pbuf` behind the curtains I this that `ipc_spsc_pbuf` is a proper name for that. My 2 cents.

**Comment by carlocaione** [comment]
> What's the problem with leaving this `icmsg`?

Because especially in this case the IPC service is really a wrapper around `spsc_pbuf` so why using two different names? It's just confusing for the user.

File: tests/lib/spsc_pbuf/src/main.c (hunk 1 above)
Comment (nordic-krch) [reply tests/lib/spsc_pbuf/src/main.c:20]: done, changed to `pb`

File: subsys/ipc/ipc_service/backends/Kconfig [hunk 2]
```diff
@@ -22,7 +22,7 @@ config IPC_SERVICE_BACKEND_ICMSG
 	bool "ICMSG backend with ICMsg buffer"
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply subsys/ipc/ipc_service/backends/Kconfig]: done

**Comment by nordic-krch** [comment]
I'm a bit reluctant to change the name of the backend as there is no consensus.

File: doc/kernel/data_structures/spsc_pbuf.rst [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+Single Producer Single Consumer Packet Buffer
+=============================================
+
+A :dfn:`Single Producer Single Consumer Packet Buffer (SPSC_PBUF)` is a circular
+buffer, whose contents are stored in first-in-first-out order. Variable size
+packets are stored in the buffer. Packet buffer works under assumption that there
+is a single context that produces packages and a single context that consumes the
+data.
+
+Implementation is focused on perfomance and memory footprint.
```
Comment (dcpleung) [inline doc/kernel/data_structures/spsc_pbuf.rst:12]: s/perfomance/performance/ (missing 'r')

File: doc/kernel/data_structures/spsc_pbuf.rst [hunk 4]
```diff
@@ -0,0 +1,17 @@
+.. _spsc_pbuf:
+
+Single Producer Single Consumer Packet Buffer
+=============================================
+
+A :dfn:`Single Producer Single Consumer Packet Buffer (SPSC_PBUF)` is a circular
+buffer, whose contents are stored in first-in-first-out order. Variable size
+packets are stored in the buffer. Packet buffer works under assumption that there
+is a single context that produces packages and a single context that consumes the
```
Comment (dcpleung) [inline doc/kernel/data_structures/spsc_pbuf.rst:9]: Maybe `packets` instead of `packages`?

**Review by pdunaj — DISMISSED** [review]

**Comment by pdunaj** [comment]
> I'm a bit reluctant to change the name of the backend as there is no consensus.

I am personally against it. The current name is as good as any other and the backend is already recognized by its name in various places. Name change will lead to additional maintenance work giving absolutely no benefits.

**Review by eobalski — CHANGES_REQUESTED** [review]
Please change to `project(spsc_pbuf)` https://github.com/zephyrproject-rtos/zephyr/blob/784ed03bb061788927bd53699f376f0f41d13e1a/tests/lib/spsc_pbuf/CMakeLists.txt#L5

Other than than, its ok.

**Review by dcpleung — DISMISSED** [review]

**Comment by nordic-krch** [comment]
@carlocaione can we skip renaming of ipc backend? I would like to push it forward.

**Comment by carlocaione** [comment]
> @carlocaione can we skip renaming of ipc backend? I would like to push it forward.

yup, no problem.

**Review by carlocaione — APPROVED** [review]

**Comment by nordic-krch** [comment]
@dcpleung @pdunaj can you reapprove. No changes, only upmerge.
@emob-nordic can you take a final look. Your last comment got applied.

**Review by pdunaj — APPROVED** [review]

**Review by eobalski — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Pull request merged**